browser/ui/commander/fuzzy_finder.cc

```diff
@@ -0,0 +1,455 @@
+    matched_ranges->clear();
+    return 0;
+  }
+  if (match_length > 0) {
+    DCHECK(match_start != -1);
+    matches.emplace_back(match_start, h_iter.array_pos(), match_length,
+                         match_began_on_boundary, gap_size_before_match);
+  }
+  for (const MatchRecord& match : matches) {
+    matched_ranges->push_back(match.range);
+  }
+  double score = ScoreForMatches(matches, LengthInCodePoints(needle),
+                                 LengthInCodePoints(haystack));
+  score *= kPrefixScore;  // Normalize so that a prefix always wins.
+  return score;
+}
+// Converts a list of indices in `positions` into contiguous ranges and fills
+// `matched_ranges` with the result.
+// For example: [0, 1, 4, 7, 8, 9] -> [{0, 2}, {4, 1}, {7, 3}].
+void ConvertPositionsToRanges(const std::vector<size_t>& positions,
+                              std::vector<gfx::Range>* matched_ranges) {
+  size_t n = positions.size();
+  DCHECK(n > 0);
+  size_t start = positions.front();
+  size_t length = 1;
+  for (size_t i = 0; i < n - 1; ++i) {
+    if (positions.at(i) + 1 < positions.at(i + 1)) {
+      // Noncontiguous positions -> close out the range.
+      matched_ranges->emplace_back(start, start + length);
+      start = positions.at(i + 1);
+      length = 1;
+    } else {
+      ++length;
+    }
+  }
+  matched_ranges->emplace_back(start, start + length);
+}
+
+// Returns the maximum score for the given matrix, then backtracks to fill in
+// `matched_ranges`. See fuzzy_finder.md for extended discussion.
+int ScoreForMatrix(const std::vector<int> score_matrix,
+                   size_t width,
+                   size_t height,
+                   const std::vector<size_t> codepoint_to_offset,
+                   std::vector<gfx::Range>* matched_ranges) {
+  // Find winning score and its index.
+  size_t max_index = 0;
+  int max_score = 0;
+  for (size_t i = 0; i < width; i++) {
+    int score = score_matrix[(height - 1) * width + i];
+    if (score > max_score) {
+      max_score = score;
+      max_index = i;
+    }
+  }
+
+  // Backtrack through the matrix to find matching positions.
+  std::vector<size_t> positions = {codepoint_to_offset[max_index]};
+  size_t cur_i = max_index;
+  size_t cur_j = height - 1;
+  while (cur_j > 0) {
+    // Move diagonally...
+    --cur_i;
+    --cur_j;
+    // ...then scan left until the score stops increasing.
+    int current = score_matrix[cur_j * width + cur_i];
+    int left = cur_i == 0 ? 0 : score_matrix[cur_j * width + cur_i - 1];
+    while (current < left) {
+      cur_i -= 1;
+      if (cur_i == 0) {
+        break;
+      }
+      current = left;
+      left = score_matrix[cur_j * width + cur_i - 1];
+    }
+    positions.push_back(codepoint_to_offset[cur_i]);
+  }
+
+  base::ranges::reverse(positions);
+  ConvertPositionsToRanges(positions, matched_ranges);
+  return max_score;
+}
+
+}  // namespace
+
+namespace commander {
+
+FuzzyFinder::FuzzyFinder(const std::u16string& needle)
```

Comment:
there's a lot of stuff going on here. Hopefully this will not be a support burden.

Reply:
cc @fallaciousreasoning.